### PR TITLE
beam_makeops: Make syntax nicer

### PR DESCRIPTION
**Pull request by bjorng**
This pull request updates the syntax of `*..tab` files to make it nicer and more explicit.

Repeating variables in patterns is no longer allowed. For example:

```
move R R =>
```

This must now be written:

```
move R1 R2 | equal(R1, R2) => _
```

This examples also illustrates another change. When instructions are to be removed (rather than replaced), the substitution on the right-hand side must now be written as '_'. That makes it clearer, and also enabling automatic continuation of lines that end in `|` or `=>`. For example, a complex transformation such as:

```
is_tuple NotTupleFail Src | \
  is_tagged_tuple WrongRecordFail Tuple Arity Atom | \
  equal(Src, Tuple) => \
   i_is_tagged_tuple_ff NotTupleFail WrongRecordFail Src Arity Atom | \
   current_tuple Src
```

can now be written simply as:

```
is_tuple NotTupleFail Src |
  is_tagged_tuple WrongRecordFail Tuple Arity Atom |
  equal(Src, Tuple) =>
   i_is_tagged_tuple_ff NotTupleFail WrongRecordFail Src Arity Atom |
   current_tuple Src
```

### CODE REVIEW

File: erts/emulator/beam/emu/ops.tab [hunk 1]
```diff
@@ -479,15 +502,16 @@ send
 # Optimized comparisons with one immediate/literal operand.
 #
 
-is_eq_exact Lbl S S =>
+is_eq_exact Lbl S S2 | equal(S, S2) => _
```
Comment (jhogberg) [inline erts/emulator/beam/emu/ops.tab:505]: Nitpick (same for ARM/x86):
```suggestion
is_eq_exact Lbl LHS RHS | equal(LHS, RHS) => _
```

File: erts/emulator/beam/emu/ops.tab [hunk 2]
```diff
@@ -559,7 +583,7 @@ put_tuple2 xy I *
 #
 put_list Const=c n Dst => move Const x | put_list x n Dst
 
-put_list Src Dst=x Dst => update_list Src Dst
+put_list Src Dst=x Dst2 | equal(Dst, Dst2) => update_list Src Dst
```
Comment (jhogberg) [inline erts/emulator/beam/emu/ops.tab:586]: Nitpick:
```suggestion
put_list Head Tail=x Dst | equal(Tail, Dst) => update_list Head Dst
```

File: erts/emulator/beam/jit/arm/instr_common.cpp [hunk 3]
```diff
@@ -1079,12 +1079,8 @@ void BeamModuleAssembler::emit_i_is_tuple(const ArgVal &Fail,
...
         comment("skipped header test since we know it's a tuple when boxed");
 
-        /* Note: ARG1 will NOT be set. This is OK since the operand
-         * for `current_tuple` has a type; that operand will not match
-         * the type-less operand for `get_tuple_element`. Thus, there
-         * will always be a `load_tuple_ptr` instruction emitted if
-         * this instruction is immediately followed by a
-         * `get_tuple_element` instruction. */
+        /* Must still set ARG1. */
+        emit_untag_ptr(ARG1, src.reg);
```
Comment (jhogberg) [inline erts/emulator/beam/jit/arm/instr_common.cpp:1083]: This is identical on both branches, maybe hoist it?

File: erts/emulator/beam/jit/x86/instr_common.cpp [hunk 4]
```diff
@@ -1187,15 +1187,11 @@ void BeamModuleAssembler::emit_i_is_tuple(const ArgVal &Fail,
...
-
-        /* Note: ARG2 will NOT be set. This is OK since the operand
-         * for `current_tuple` has a type; that operand will not match
-         * the type-less operand for `get_tuple_element`. Thus, there
-         * will always be a `load_tuple_ptr` instruction emitted if
-         * this instruction is immediately followed by a
-         * `get_tuple_element` instruction. */
+        /* We must be careful to still leave the pointer to the tuple
+         * in ARG2. */
+        mov_arg(ARG2, Src);
```
Comment (jhogberg) [inline erts/emulator/beam/jit/x86/instr_common.cpp:1192]: Maybe hoist this move, and change the comment to explain why we're going through `ARG2`?

File: erts/emulator/beam/jit/x86/instr_common.cpp [hunk 5]
```diff
@@ -1187,15 +1187,11 @@ void BeamModuleAssembler::emit_i_is_tuple(const ArgVal &Fail,
...
+        (void)emit_ptr_val(ARG2, ARG2);
+        a.test(ARG2.r8(), imm(_TAG_PRIMARY_MASK - TAG_PRIMARY_BOXED));
```
Comment (jhogberg) [inline erts/emulator/beam/jit/x86/instr_common.cpp:1194]: Swapping these improves latency a teeny tiny bit:
```suggestion
        a.test(ARG2.r8(), imm(_TAG_PRIMARY_MASK - TAG_PRIMARY_BOXED));
        (void)emit_ptr_val(ARG2, ARG2);
```

File: erts/emulator/beam/jit/x86/ops.tab [hunk 6]
```diff
@@ -432,8 +456,13 @@ put_tuple2 S A *
 
 store_cons/2
 put_list Hd Tl Dst => put_cons Hd Tl | store_cons u=1 Dst
-store_cons Len Dst | put_list Hd Dst Dst | distinct(Dst, Hd) => combine_conses(Len, Dst, Hd)
-store_cons Len Dst | put_list Hd Dst OtherDst => store_cons Len Dst | append_cons u Hd | store_cons u=1 OtherDst
+
+store_cons Len Dst | put_list Hd Dst2 Dst3 |
+  equal(Dst, Dst2) | equal(Dst, Dst3) | distinct(Dst, Hd) =>
+    combine_conses(Len, Dst, Hd)
```
Comment (jhogberg) [inline erts/emulator/beam/jit/x86/ops.tab:462]: Nitpick:
```suggestion
store_cons Len Dst | put_list Head Tail Dst2 |
  equal(Dst, Tail) | equal(Dst, Dst2) | distinct(Dst, Head)  =>
    combine_conses(Len, Dst, Head)
```
(same on the lines below)

File: erts/emulator/beam/predicates.tab [hunk 7]
```diff
@@ -215,27 +215,33 @@ pred.map_key_sort(Size, Rest) {
...
+pred.distinct(Val1, Val2) {
+    if (Val1.type != Val2.type) {
         return 1;
-    } else if (Reg1.type == TAG_x || Reg1.type == TAG_y) {
+    } else if (Val1.type == TAG_x || Val1.type == TAG_y) {
         /* We must not compare the type indices (if any). */
-        return (Reg1.val & REG_MASK) != (Reg2.val & REG_MASK);
+        return (Val1.val & REG_MASK) != (Val2.val & REG_MASK);
+    } else if (Val1.type == TAG_n) {
+        /* NIL has no associated value. */
```
Comment (jhogberg) [inline erts/emulator/beam/predicates.tab:226]: Do you think it would be better if we normalized the value of `TAG_n` in `beam_file` instead?

**Review by jhogberg — APPROVED** [review]

File: erts/emulator/beam/predicates.tab (hunk 7 above)
Comment (bjorng) [reply erts/emulator/beam/predicates.tab:226]: Yes, will do.

File: erts/emulator/beam/jit/arm/instr_common.cpp (hunk 3 above)
Comment (bjorng) [reply erts/emulator/beam/jit/arm/instr_common.cpp:1083]: Will do.

File: erts/emulator/beam/jit/x86/instr_common.cpp (hunk 5 above)
Comment (bjorng) [reply erts/emulator/beam/jit/x86/instr_common.cpp]: But `emit_ptr_val()` will do an `and` when literals are tagged (always on Windows), which will disturb ZF.

File: erts/emulator/beam/predicates.tab (hunk 7 above)
Comment (bjorng) [reply erts/emulator/beam/predicates.tab:226]: I take it back. I think it's safeer as it is. `beam_file` already sets the value to NIL, but any generator could set the type to `TAG_n` and forget to set the value (or set it to a different value). In fact, [beam_file itself](https://github.com/erlang/otp/blob/07b78d31b1d9edb78aa4a089cc733634ff06e541/erts/emulator/beam/beam_file.c#L1755) sets the type to `TAG_n` without setting the value.

**Review by garazdawi — APPROVED** [review]

**Pull request merged**